File: bench/java/binarytrees.java
// Binary trees, built and consumed, three repetitions of growing depth. The
// allocation kernel: every node is one heap object in every language, so the
// kernel measures the allocator and the garbage collector.
//
// A leaf carries the path tag it was built under, so the checksum is a real
// function of the shape rather than a node count.

public class binarytrees {
    static final int FIRST_DEPTH = 13;
    static final int REPETITIONS = 3;

    static class Tree {
        Tree left;  // null for a leaf
        Tree right; // null for a leaf
        long value; // path tag, meaningful for a leaf
    }

    static Tree build(int depth, long tag) {
        Tree node = new Tree();
        if (depth == 0) {
            node.value = tag;
            return node;
        }
        node.left = build(depth - 1, tag * 2);
        node.right = build(depth - 1, tag * 2 + 1);
        return node;
    }

    static long sumTree(Tree tree) {
        if (tree.left == null)
            return tree.value;
        return sumTree(tree.left) + sumTree(tree.right);
    }

    public static void main(String[] args) {
        long total = 0;
        for (int rep = 0; rep < REPETITIONS; rep++) {
            total += sumTree(build(FIRST_DEPTH + rep, 1));
        }
        System.out.println(total);
    }
}

File: bench/rust/src/bin/binarytrees.rs
// Binary trees, built and consumed, three repetitions of growing depth. The
// allocation kernel: every node is one heap block in every language, so the
// kernel measures the allocator and the drop path -- malloc/free, a garbage
// collector, or Box's drop glue.
//
// A leaf carries the path tag it was built under, so the checksum is a real
// function of the shape rather than a node count.

const FIRST_DEPTH: i64 = 13;
const REPETITIONS: i64 = 3;

enum Tree {
    Leaf(i64),
    Node(Box<Tree>, Box<Tree>),
}

fn build(depth: i64, tag: i64) -> Tree {
    if depth == 0 {
        Tree::Leaf(tag)
    } else {
        Tree::Node(
            Box::new(build(depth - 1, tag * 2)),
            Box::new(build(depth - 1, tag * 2 + 1)),
        )
    }
}

fn sum_tree(tree: Tree) -> i64 {
    match tree {
        Tree::Leaf(value) => value,
        Tree::Node(left, right) => sum_tree(*left) + sum_tree(*right),
    }
}

fn main() {
    let mut total = 0;
    for rep in 0..REPETITIONS {
        let tree = build(FIRST_DEPTH + rep, 1);
        total += sum_tree(tree);
    }
    println!("{}", total);
}
